librz/include/rz_util/set.h

```diff
@@ -24,10 +24,33 @@ typedef HtUP SetU;
 
 RZ_API RZ_OWN SetU *set_u_new(void);
 RZ_API void set_u_add(RZ_NONNULL SetU *set, ut64 u);
+RZ_API ut64 set_u_size(SetU *s);
 RZ_API bool set_u_contains(RZ_NONNULL SetU *set, ut64 u);
 RZ_API void set_u_delete(RZ_NONNULL SetU *set, ut64 u);
 RZ_API void set_u_free(RZ_NULLABLE SetU *set);
 
+typedef struct {
+	st64 ti; ///< Table index
+	ut64 bi; ///< Bucket index
+	ut64 v; ///< Current value of the iteration
+} SetUIter;
+
+RZ_API void advance_set_u_iter(SetU *s, SetUIter *it);
+
+#define set_u_iter_reset(iter) \
+	do { \
+		iter.ti = 0; \
+		iter.bi = 0; \
+		iter.v = 0; \
+	} while (0)
+/**
+ * The adcvance_set_u_iter() sets iter.ti always to the entry of the next table to check.
+ * So our condition checks if ti <= set->size.
+ */
+#define set_u_foreach(set, iter) \
+	if (set && set_u_size(set) > 0) \
+		for (advance_set_u_iter(set, &iter); iter.ti <= set->size; advance_set_u_iter(set, &iter))
```

Comment:
Just wondering can it be done without `set_u_iter_reset`?
```suggestion
	if (set && set_u_size(set) > 0) \
		for (iter.ti = 0, iter.bi = 0, iter.v = 0, advance_set_u_iter(set, &iter); iter.ti <= set->size; advance_set_u_iter(set, &iter))
```

Reply:
> Just wondering can it be done without `set_u_iter_reset`?

The `set_u_iter_reset` is only there, so the same iter can be used twice in a function, without the need to init another one.
Also it is supposed to signal, that the `iter` can not just be reused without any extra steps (resetting it).

I tried in the beginning to implement it with a single pointer for value `v`, but failed.
I think the problem was, that the logic of `bi/ti` index tracking of `advance_set_u_iter()`, doesn't work in the for loop alone.

Coming up with this version already took a while. But then again, I am not a C magician. So maybe there is a more elegant solution.

Reply:
So have you tried the suggested one?

Reply:
Ah, where is my head these days -.- Of course it works. Thanks for the suggestion.